[IMP] hw_drivers: allow iotbox to be used as a kiosk

loouis-t opened the pull request:
- Moved browser logic inside of a class (Browser),
- Added screen rotation logic,
- Added support for touchscreens: 
    - Screen matrix sync with screen orientation, 
    - Cursor hiding in kiosk, 
    - Disabled zoom in kiosk,
- Added screen orientation and browser url configuration saving logic.

Enterprise PR: [https://github.com/odoo/enterprise/pull/67092](https://github.com/odoo/enterprise/pull/67092)
Task: 3598744

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/174009.png)](https://mergebot.odoo.com/odoo/odoo/pull/174009)

vlst-odoo commented on addons/hw_drivers/iot_handlers/drivers/DisplayDriver_L.py:
```diff
@@ -85,32 +99,62 @@ def load_url(self):
-            return "xdotool threw an error, maybe it is not installed on the IoTBox"
+        return url
+
+    def update_customer_facing_display(self, origin, html=None):
+        if origin == self.owner:
+            self.rendered_html = html
+            self.event_data.set()
+
+    def get_serialized_order(self):
+        # IMPLEMENTATION OF LONGPOLLING
+        # Times out 2 seconds before the JS request does
+        if self.event_data.wait(28):
+            self.event_data.clear()
+            return {'rendered_html': self.rendered_html}
+        return {'rendered_html': False}
```
i think that this might be a rebase conflict. This looks like old code

vlst-odoo commented on addons/hw_drivers/iot_handlers/drivers/DisplayDriver_L.py:
```diff
@@ -173,3 +217,42 @@ def get_iot_devices(self):
 
         return json.dumps({'iot_device_status': iot_device})
+
+    @http.route('/kiosk/display', type='json', auth='none', methods=['POST'])
```
why do you need a new controller for this ? Why not simply point the kiosk to the regular address of the self order page ?

vlst-odoo commented on addons/hw_drivers/iot_handlers/drivers/DisplayDriver_L.py:
```diff
@@ -173,3 +217,42 @@ def get_iot_devices(self):
+
+        uri = f'{helpers.get_odoo_server_url()}/pos-self/{pos_id}?access_token={access_token}'
+
+        display.set_orientation(Orientation.RIGHT)  # default orientation for kiosk mode
```
you can call the method w\o a parameter, given that you declare the default value already

loouis-t replied:
it's the default orientation for kiosk mode, not the default param of the method: we want the screen to be landscape 90% of the time, but when in kiosk mode, the new default is portrait (`Orientation.RIGHT`)

qle-odoo commented on addons/hw_drivers/tools/helpers.py:
```diff
@@ -262,12 +262,15 @@ def get_ssid():
             return False
-    return read_file_first_line('odoo-remote-server.conf')
+
+    return read_file_first_line('/home/pi/odoo-remote-server.conf', absolute=True)
```
why ?
Path.home() not return '/home/pi' ?

loouis-t replied:
did not always, depending on `os.environ['HOME']`, which changes in `browser.py` to `/tmp/0` for chromium

qle-odoo commented on addons/hw_drivers/tools/helpers.py:
```diff
@@ -439,12 +442,14 @@ def path_file(filename):
-    path = path_file(filename)
+
+def read_file_first_line(filename, absolute=False):
+    path = Path(filename) if absolute else file_path(filename)
```
?

loouis-t replied:
`file_path` is not made for absolute paths

loouis-t commented:
@odoo/rd-security 
could you have a look at this?
We basically moved the whole

mart-e commented:
@loouis-t but why is the whole method safe?

Ysoroko commented on addons/hw_drivers/browser.py:
```diff
@@ -0,0 +1,118 @@
+            [
+                self.browser,
+                self.url,
+                '--disk-cache-dir=/dev/null',
+                '--disk-cache-size=1',
+                '--log-level=3',
```
this could be better with some explanation comments

Ysoroko commented on addons/hw_drivers/browser.py:
```diff
@@ -0,0 +1,118 @@
+    def close_browser(self):
+        """close the browser"""
+        subprocess.run(['pkill', self.browser.split('-')[0] + '*'], check=False)
+        subprocess.run(['rm', '-rf', f'/tmp/{self._x_screen}/.cache'], check=False)  # Remove cached files to free space
```
is this really necessary or avoidable if we configure the browser properly?

loouis-t commented:
> @loouis-t but why is the whole method safe?

## Environment 

- The browser is still running in `/tmp` on the IoT Box: we then copied the environment variables `return os.environ.copy()` and updated the copied object to give it to the `subprocess` call

```python
for key in ['HOME', 'XDG_RUNTIME_DIR', 'XDG_CACHE_HOME']:
    self.env[key] = '/tmp/' + self._x_screen
```

## Subprocess

- Open Chromium on the right url and with the right params (e.g. `--kiosk` to avoid being able to leave the window on a touchscreen)

```python
# addons/hw_drivers/browser.py
self.instance = subprocess.Popen(
    [
        self.browser,
        self.url,
        '--disk-cache-dir=/dev/null',
        '--disk-cache-size=1',
        '--log-level=3',
        *self.chromium_additional_args,
   ],
   env=self.env,
)
```

- Kill chromium process (browser name is hard coded)

```python
# addons/hw_drivers/browser.py
subprocess.run(['pkill', self.browser.split('-')[0] + '*'], check=False)
```

- Handle screen rotation using `xrandr` (which would just fail if `orientation.value` differed from `xrandr` allowed options) 

```python
# addons/hw_drivers/iot_handlers/drivers/DisplayDriver_L.py
subprocess.run(['xrandr', '-o', orientation.value], check=True)
```

- Handle touchscreen rotation as well

```python
# addons/hw_drivers/iot_handlers/drivers/DisplayDriver_L.py
subprocess.run([file_path('hw_drivers/tools/sync_touchscreen.sh'), str(int(self._x_screen) + 1)], check=False)
```

This script will look for touchscreens, get their `xinput` ids, their `xrandr` names, and map the touchscreen to the screen's orientation with `xinput map-to-output`. 

The only possible input here is the screen's number (e.g. `:0`, `:0.1`, etc.), which is either hard coded or read by the `DisplayInterface`.

loouis-t commented:
@mart-e just ping you because I'm not sure the previous did

loouis-t commented on addons/hw_drivers/browser.py:
```diff
@@ -0,0 +1,129 @@
+        :param kiosk: Whether the browser should be in kiosk mode
+        """
+        self.url = url
+        self.browser = 'chromium-browser'
```
this would require using a complete lib, overkill in our case

mart-e commented on addons/hw_drivers/iot_handlers/drivers/DisplayDriver_L.py:
```diff
@@ -86,52 +94,68 @@ def load_url(self):
+    def customer_facing_display(self, action, pos_id=None, access_token=None, data=None):
+        display = self.ensure_display()
+        if action in ['open', 'open_kiosk']:
             origin = helpers.get_odoo_server_url() or http.request.httprequest.referrer.split('://')[-1].split('/')[0]
```
so assuming you can’t trust the referer, it would be a way to inject malicious url in the configuration file for the next person?
update_url -> open_browser -> save_browser_state -> write_file('browser-url.conf', url)

loouis-t replied:
@mart-e now that I removed the possibility to get the origin from the referrer, the url can only be the db's one.

Ysoroko commented on addons/hw_drivers/tools/sync_touchscreen.sh:
```diff
@@ -0,0 +1,26 @@
+output_name=$(xrandr --listactivemonitors | awk '{print $NF}' | tail -n +2 | awk -v screen="$screen" 'NR==screen')
+
+# Map touchscreen matrix to screen configuration (orientation)
+xinput map-to-output $device_id_0 "$output_name"
+xinput map-to-output $device_id_1 "$output_name"
```
note: xinput is not currently present in the new IoT Box Image PR, added in this PR: https://github.com/odoo/odoo/pull/176751

Ysoroko commented:
Note: this PR needs to wait for the following PRs to be merged in master before:
1) https://github.com/odoo/odoo/pull/176751
2) https://github.com/odoo/odoo/pull/176645

It also requires the new IoT Box image to work

Ysoroko commented:
Tested with Elo Touch kiosk + the new image + new DisplayInterface_L.py + the other files from this PR (browser.py) etc.
Everything works
1) Opening the Kiosk on the IoT Box ✅
2) Changing the display orientation ✅
3) The "kiosk" url and mode are stored on the db and after reboot the browser opens the kiosk url in kiosk mode ✅
4) The screen matrix and touchscreen settings work ✅

Ysoroko commented:
@robodoo r+
